tritium-metrics/src/main/java/com/palantir/tritium/event/metrics/TaggedMetricsServiceInvocationEventHandler.java

```diff
@@ -0,0 +1,117 @@
+import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An implementation of {@link AbstractInvocationEventHandler} whose purpose is to provide tagged metrics for
+ * classes which look like services.
+ *
+ * Specifically, this class will generate metrics with the following parameters:
+ *
+ * <ul>
+ *     <li>Metric Name: the service name supplied to the constructor</li>
+ *     <li>Tag - service-name: The simple name of the invoked class</li>
+ *     <li>Tag - endpoint: The name of the method that was invoked</li>
+ *     <li>Tag - cause: When an error is hit, this will be filled with the full class name of the cause.</li>
+ * </ul>
+ */
+public class TaggedMetricsServiceInvocationEventHandler extends AbstractInvocationEventHandler<InvocationContext> {
+
+    private static final Logger logger = LoggerFactory.getLogger(TaggedMetricsServiceInvocationEventHandler.class);
+
+    private final TaggedMetricRegistry taggedMetricRegistry;
+    private final String serviceName;
+
+    public TaggedMetricsServiceInvocationEventHandler(
+            TaggedMetricRegistry taggedMetricRegistry,
+            String serviceName) {
+        super(getEnabledSupplier(serviceName));
+        this.taggedMetricRegistry = checkNotNull(taggedMetricRegistry, "metricRegistry");
+        this.serviceName = checkNotNull(serviceName, "serviceName");
+    }
+
+    private static String failuresMetricName() {
+        return "failures";
+    }
+
+    private static BooleanSupplier getEnabledSupplier(final String serviceName) {
+        return InstrumentationProperties.getSystemPropertySupplier(serviceName);
+    }
+
+    @Override
+    public final InvocationContext preInvocation(Object instance, Method method, Object[] args) {
+        return DefaultInvocationContext.of(instance, method, args);
+    }
+
+    @Override
+    public final void onSuccess(@Nullable InvocationContext context, @Nullable Object result) {
+        if (context == null) {
+            logger.debug("Encountered null metric context likely due to exception in preInvocation");
+            return;
+        }
+
+        long nanos = System.nanoTime() - context.getStartTimeNanos();
+        MetricName finalMetricName = MetricName.builder()
+                .safeName(serviceName)
+                .putSafeTags("service-name", context.getMethod().getDeclaringClass().getSimpleName())
+                .putSafeTags("endpoint", context.getMethod().getName())
+                .build();
+        taggedMetricRegistry.timer(finalMetricName)
+                .update(nanos, TimeUnit.NANOSECONDS);
+    }
+
+    @Override
+    public final void onFailure(@Nullable InvocationContext context, @Nonnull Throwable cause) {
+        markGlobalFailure();
+        if (context == null) {
+            logger.debug(
+                    "Encountered null metric context likely due to exception in preInvocation: {}",
+                    UnsafeArg.of("exception", cause),
```

Comment:
this is unnecessary, we're providing the full Throwable in the next arg

Reply:
I'll make the fix to the repo where I copied this....